Support analytic data

nilsdeppe opened the pull request:
## Proposed changes

- Support analytic data for evolutions.

### Types of changes:

- [ ] Bugfix
- [x] New feature

### Component:

- [ ] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] The PR passes all checks, including unit tests, `clang-tidy` and `IWYU`. For
  instructions on how to perform the CI checks locally refer to the [Dev guide
  on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run `make doc`
  to generate the documentation locally into `BUILD_DIR/docs/html`. Then open
  `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

nilsdeppe committed:
Add support for loading analytic data to GRMHD Div Clean

nilsdeppe committed:
Add GRMHD executable that loads analytic data

moxcodes commented on src/Evolution/Systems/GrMhd/ValenciaDivClean/Initialize.hpp:
```diff
@@ -66,14 +80,34 @@ struct Initialize {
+            return Parallel::get<analytic_data_tag>(local_cache)
+                .equation_of_state();
+          })(
+          std::integral_constant<
```
the first argument can be`detail::has_analytic_solution_alias<Metavariables>`, which is either `true_type` or `false_type` (and below). It seems these are the only places the `_v` version is used, so that can probably be omitted as well.

moxcodes commented on src/Evolution/Executables/GrMhd/ValenciaDivClean/EvolveValenciaDivCleanBlastWaveFwd.hpp:
```diff
@@ -0,0 +1,6 @@
+
+#pragma once
+
+struct EvolutionMetavars;
```
For our understanding, why is this file necessary? In other cases, it would seem preferable to just forward-declare the single struct.

nilsdeppe replied:
This is needed to provide the magic type alias `metavariables` globally. I currently can't remember why it's necessary, but it is needed if `EvolutionMetavars` is a template like in the scalar wave case. Does that help?

kidder replied:
It is needed for `cmake/AddSpectreExecutable.cmake` to generate the actual executable.  It is only necessary when the metavariables are a template, but since we expect most systems to be templated on dimension, it was decided to do it this way for all systems rather than support two setups.

moxcodes reviewed (commented):
Group code review with @markscheel and Rob.

hrueter commented on src/Evolution/Systems/GrMhd/ValenciaDivClean/Initialize.hpp:
```diff
@@ -66,14 +80,34 @@ struct Initialize {
-          Parallel::get<solution_tag>(cache).equation_of_state();
+      auto equation_of_state = make_overloader(
+          [ initial_time, &inertial_coords ](
+              std::true_type /*is_analytic_solution*/, const auto prim_vars,
```
could you replace `auto` with the `gsl::not_null<PrimitiveVars*>`?

hrueter commented on src/Evolution/Systems/GrMhd/ValenciaDivClean/Initialize.hpp:
```diff
@@ -66,14 +80,34 @@ struct Initialize {
+            return Parallel::get<solution_tag>(local_cache).equation_of_state();
+          },
+          [&inertial_coords](std::false_type /*is_analytic_solution*/,
+                             const auto prim_vars,
```
also replace `auto`?

hrueter commented on src/Evolution/Systems/GrMhd/ValenciaDivClean/Initialize.hpp:
```diff
@@ -113,10 +147,28 @@ struct Initialize {
-          inertial_coords, initial_time, typename GrVars::tags_list{}));
+      make_overloader(
+          [ initial_time, &inertial_coords ](
+              std::true_type /*is_analytic_solution*/, const auto local_gr_vars,
```
same here?

hrueter commented on src/Evolution/Executables/GrMhd/ValenciaDivClean/EvolveValenciaDivCleanBlastWave.hpp:
```diff
@@ -0,0 +1,173 @@
+#include "Domain/DomainCreators/RegisterDerivedWithCharm.cpp"
+#include "Domain/Tags.hpp"
+#include "ErrorHandling/FloatingPointExceptions.hpp"
+#include "Evolution/Actions/ComputeVolumeDuDt.hpp"
```
rename to `ComputeTimeDerivative`

hrueter commented on src/Evolution/Executables/GrMhd/ValenciaDivClean/EvolveValenciaDivCleanBlastWave.hpp:
```diff
@@ -0,0 +1,173 @@
+#include "Evolution/VariableFixing/Actions.hpp"
+#include "Evolution/VariableFixing/FixToAtmosphere.hpp"
+#include "Evolution/VariableFixing/Tags.hpp"
+#include "NumericalAlgorithms/DiscontinuousGalerkin/Actions/ApplyBoundaryFluxesGlobalTimeStepping.hpp"
```
rename to `ApplyFluxes`

hrueter commented on src/Evolution/Executables/GrMhd/ValenciaDivClean/EvolveValenciaDivCleanBlastWave.hpp:
```diff
@@ -0,0 +1,173 @@
+  using normal_dot_numerical_flux = OptionTags::NumericalFluxParams<
+      dg::NumericalFluxes::LocalLaxFriedrichs<system>>;
+  using limiter = OptionTags::SlopeLimiterParams<
+      SlopeLimiters::Minmod<3, system::variables_tag::tags_list>>;
```
retrieve dimension from `system::volume_dim` (also below)

Alternatively comment that the system always assumes 3 dimensions

nilsdeppe replied:
Pretty much all of GRMHD is coded this way since the system is defined only in 3D. I can add a comment.

hrueter commented on src/Evolution/Executables/GrMhd/ValenciaDivClean/EvolveValenciaDivCleanBlastWave.hpp:
```diff
@@ -0,0 +1,173 @@
+  using compute_rhs = tmpl::flatten<tmpl::list<
+      Actions::ComputeVolumeFluxes,
+      dg::Actions::SendDataForFluxes<EvolutionMetavars>,
+      Actions::ComputeVolumeSources, Actions::ComputeVolumeDuDt,
```
rename to `ComputeTimeDerivative`

hrueter commented on src/Evolution/Executables/GrMhd/ValenciaDivClean/EvolveValenciaDivCleanBlastWave.hpp:
```diff
@@ -0,0 +1,173 @@
+      Actions::ComputeVolumeFluxes,
+      dg::Actions::SendDataForFluxes<EvolutionMetavars>,
+      Actions::ComputeVolumeSources, Actions::ComputeVolumeDuDt,
+      // dg::Actions::ImposeDirichletBoundaryConditions<EvolutionMetavars>,
```
Add a comment on which boundary conditions can be used at the moment.

nilsdeppe replied:
I don't know where to add this comment. You can only use periodic BCs for analytic data right now, but that's a restriction completely outside of this file or even this evolution system.

geoffrey4444 replied:
So what happens when I run the blast wave with x,y not set to be periodic?

hrueter commented on src/Evolution/Executables/GrMhd/ValenciaDivClean/EvolveValenciaDivCleanBlastWave.hpp:
```diff
@@ -0,0 +1,173 @@
+      // dg::Actions::ImposeDirichletBoundaryConditions<EvolutionMetavars>,
+      dg::Actions::ReceiveDataForFluxes<EvolutionMetavars>,
+      tmpl::conditional_t<local_time_stepping, tmpl::list<>,
+                          dg::Actions::ApplyBoundaryFluxesGlobalTimeStepping>,
```
rename to `ApplyFluxes`

hrueter commented on src/Evolution/Executables/GrMhd/ValenciaDivClean/EvolveValenciaDivCleanBlastWave.hpp:
```diff
@@ -0,0 +1,173 @@
+          compute_rhs, update_variables, Actions::Goto<EvolvePhaseStart>>>>>;
+
+  using const_global_cache_tag_list =
+      tmpl::list<analytic_data_tag,
```
This should probably be in `Initialize::const_global_cache_tag_list` similar to how `normal_dot_numerical_fluxes_tag` is handled.
But `DgElementArray` does not read that list from the initialize action. So maybe we should open an issue for this.

nilsdeppe replied:
I filed issue #1191. Please feel free to comment there if my description of what should be done is insufficient :)

hrueter commented on src/Evolution/Executables/GrMhd/ValenciaDivClean/EvolveValenciaDivCleanBlastWave.hpp:
```diff
@@ -0,0 +1,173 @@
+  using domain_creator_tag = OptionTags::DomainCreator<3, Frame::Inertial>;
+
+  static constexpr OptionString help{
+      "Evolve analytic data using the Valencia formulation of the GRMHD system "
```
List the choices of cylindrical blast wave, Local Lax Friedrichs etc. here.

hrueter reviewed (changes requested):
group review by @nilsleiffischer @hrueter

geoffrey4444 reviewed (dismissed):
Does this blast wave executable differ from my modified executable in https://github.com/geoffrey4444/spectre/commit/2adab7b43318b70c10773bbda72e713b43aeef82
? Just curious, not easy to diff on my phone...

nilsdeppe committed:
Make return types of actions explicit

This will be necessary for when we have support for forward declaring actions.

nilsdeppe commented:
- @geoffrey4444 the biggest difference is you can only apply periodic BCs
- I'm planning on adding a test for the Initialize construct.
- I'm hoping to be able to avoid travis timeouts by speeding up the compilation a bit.

Because of the last 2 points I'm marking this as "in progress".

geoffrey4444 commented:
Thanks @nilsdeppe! Sounds good. Just wondering: why restrict to periodic BCs?

nilsdeppe commented:
Because we don't have anything else implemented. The dirichlet BCs use the analytic solution, which we don't have if we are using analytic data.

hrueter commented:
Comment by @nilsleiffischer and @hrueter 
In the current code base the executables are not named after specific solutions. Therefore the executable should be renamed to something like
`EvolveValenciaDivCleanAnalyticData`.

nilsdeppe commented:
> Comment by @nilsleiffischer and @hrueter
> In the current code base the executables are not named after specific solutions. Therefore the executable should be renamed to something like
> `EvolveValenciaDivCleanAnalyticData`.

Sounds good to me.

nilsdeppe committed:
Add missing library dep to HelloWorldNoCharm

nilsdeppe commented:
@hrueter @nilsleiffischer @moxcodes I addressed the comments. This will timeout on the builds until #1216 is merged but is otherwise ready for review.

hrueter commented on src/Evolution/Executables/GrMhd/ValenciaDivClean/EvolveValenciaDivCleanAnalyticData.hpp:
```diff
@@ -0,0 +1,177 @@
+      "Local time-stepping: none\n"
+      "Boundary conditions: only periodic are currently supported\n"};
+
+  enum class Phase { Initialization, Evolve, Exit };
```
Do you have an observe action ready for this, which does not use an analytic solution? If so, can you add it here or in another PR?

nilsdeppe replied:
I do not have such an action written

nilsvu replied:
It‘ll presumably be just the current grmhd observer with everything that needs an analytic solution deleted, to start with. Out of scope for this PR I suppose, but will be necessary for @geoffrey4444 to rebase the blast wave.

nilsdeppe replied:
Yes, I'm not sure what he is observing. I think often you don't actually want to observe all the variables anyway because it's a lot of data for things you might not care about.

hrueter reviewed (commented):
group review by @nilsleiffischer @hrueter 
squash the fixup

kidder reviewed (commented):
squash

nilsdeppe commented:
Okay, squashed. I'd like to rebase this on #1216 before we merge to make sure Travis is working correctly

kidder commented:
@nilsdeppe you can rebase

nilsdeppe commented:
Rebased. :crossed_fingers:

nilsvu commented:
@nilsdeppe are you still planning to add a test for the initializer?

nilsdeppe commented:
@nilsleiffischer I'd like to defer that so I can focus on implementing more generic boundary conditions first. All the new initialization code needs tests so to do it right will be a fair amount of work.

nilsvu commented:
yes I know, see #1207 😄 I agree that testing the initialization is out of scope for this PR.

kidder commented:
@wthrowe please look at
